feat: [MCA-1028] Rework the stack splitting functionality

nikolauska opened the pull request:
BREAKING CHANGE: New splitting is not backwards compatible

With the splitting changes the way substacks are created is
changed. This means to deploy this version of monitoring
old monitoring stack will need to be removed first

Commit:
feat: Rework the chunking logic to better calculate resources

BREAKING CHANGE: New chunking is not bacwards compatible

With the chunking changes the way substacks are created is
changed. This means to deploy this version of monitoring
old monitoring stack will need to be removed first

Commit:
chore: Remove metric from resource count

Commit:
chore: Fix old calculate test value

Commit:
chore: Fix alarm enabled check

biphobe commented on src/utils/config.ts:
```diff
@@ -439,34 +439,108 @@ export function configAutoResolve(
+/**
+ * Generate chunked array of the metric alarms
+ */
+export function chunkByStackLimit<T extends ConfigMetricAlarms = ConfigMetricAlarms>(
```
It would be a good idea to use verb in the naming here to make it more meaningful. Naming could be improved in general, e.g. `results` doesn't say much and could be used to describe basically anything, `calculateResources` doesn't indicate that it returns anything and doesn't tell what it actually calculates etc.

nikolauska replied:
Naming has now been improved to `resourceCountMap` and the function was also changed to `getResourceCountMap`

biphobe commented on src/utils/config.ts:
```diff
@@ -439,34 +439,108 @@ export function configAutoResolve(
+  let currentCount = 0;
+  let index = 0;
+  for (const [key, [count, metrics]] of Object.entries(results)) {
+    if ((currentCount + count) > 500) {
```
If I'm not mistaken looking at these conditions, `CDKMetadata` is not taken into consideration despite using a resource slot in the stack resources (unless version reporting is explicitly disabled). This could result in incorrect resource count calculation.

For example - 100 lambdas with 5 alerts would generate 500 resources + `CDKMedatada`, so 501 in total, but the code would sum it as 500 and wouldn't shift the index - which would result in CDK attempting to deploy 501 resources and failing in such attempt.

nikolauska replied:
Latest change fixed this by actually finding if the version reporting is enabled from the context. It won't be included in the calculation, but it will decrease the max resource count by one if it is enabled so the `CDKMetada` can fit in to  the stack.

Commit:
chore: Include CDKMetadata in resource chunking

If version reporting is enabled then the CDKMetadata
resource is added to every lambda. With these changes
mca-monitoring can include that in to the stack splitting
calculations.
